### DISCUSSION TITLE
How to connect to Kafka cluster (that has tls encryption enabled) from a Nodejs Kubernetes Pod

### DISCUSSION THREAD

**Discussion by spoorthyredi2** [strimzi/strimzi-kafka-operator#7604] [Q&A]
Hi, 

I have tested my Strmizi Kafka cluster with ssl using my Kafka-CLI Java clients and having cert imported to cacerts and mentioning ssl config parameters in a properties file.
I have also tested my Nodejs Kubernetes Pods connect to Cluster via Plaintext Listener. 

Now, how can I do the same via Node apps running on Kubernetes Pods, but with SSL enabled listener?
The connection URL is not changing, so what difference is it going to make to the code ?  And how do we give the Cert and its password.  Any input will help

**Comment by spoorthyredi2** [comment]
This is the KafkaJS snippet for ssl . 

![image](https://user-images.githubusercontent.com/116352972/201032895-2bf5c3a0-dac5-4daf-88e2-d8176f371cb4.png)

For java client testing, I have only used the ca.crt and ca.password from the "cluster-ca-cert" secret. 
So here , it says 
ca -- This is the location to the system cacerts file
key --  is this the ca.password ? or ca.key ?
cert -- This is the cluster ca.crt file path

Is my above understanding correct

**Reply by spoorthyredi2** [reply]
So finally we tried with below values from our code
ssl: {
    key: "",
    cert: ""
    ca: strimzi-cluster-ca.crt   //ca.crt from the Cluster-ca-cert secret
 passphrase : //ca.password from cluster-ca-certsecret
  }

Got some cert errors ... and later on using one more parameter 
NODE_TLS_REJECT_UNAUTHORIZED: 0
things worked.
However we are unsure if  the connection is really TLS encrypted or  not now.   @scholzj  Please help

Is this the right way ?   Is there any other way

**Reply by spoorthyredi2** [reply]
@valdar  @jpkrohling @gunnarmorling @xiaods

**Reply by scholzj** [reply]
> However we are unsure if the connection is really TLS encrypted or not now. @scholzj Please help

Sorry, but I do not know how to help. I know nothing about KafkaJS (or Javascript). And you did not share your configuration of the Kafka cluster - so while I might understand that and tell you how you configured it, you would need to share it first.

**Reply by spoorthyredi2** [reply]
Sure.here it is ,,,, @scholzj 
  kafka:
    version: 3.2.3
    replicas: 3
    resources:
      requests:
        memory: "2Gi"
        cpu: "500m"
      limits:
        memory: "2Gi"
        cpu: "1"
    listeners:
      - name: plain
        port: 9097
        type: loadbalancer
        tls: false
      - name: plainloadb
        port: 9099
        type: loadbalancer
        tls: true
        configuration:
          bootstrap:
            annotations:
              cloud.google.com/load-balancer-type: "Internal"
    config:
      offsets.topic.replication.factor: 3
      transaction.state.log.replication.factor: 3
      transaction.state.log.min.isr: 2
      default.replication.factor: 3
      min.insync.replicas: 2
      inter.broker.protocol.version: "3.2"
    storage:

**Reply by scholzj** [reply]
If you can please format it as a code, it will be readable and it will allow to see how the whitespaces and indentation looks like which is important for reading YAMLs. You will also need to explain to which listener are you actually connecting since it looks like you have two of them there.

**Reply by spoorthyredi2** [reply]
I am trying to connect to plainloadb listner. Application is getting connected but my question is if the connection is secure or not.
![image](https://user-images.githubusercontent.com/116352972/201389210-33686584-0f02-4655-af72-a44771c11678.png)

**Reply by scholzj** [reply]
So if you connect to `plainloaddb`, then you use only TLS encryption (and server authentication, butnot TLS client authentication). So you will need to configure the `ca` option. But not the `cert` or `key` options.

**Reply by spoorthyredi2** [reply]
Yes right....that's what we used... 

1. and the value in ca option should have the same value as of the ca.crt from the cluster-ca-secret right?

**Reply by scholzj** [reply]
That should be the `ca.crt` from the `<cluster-name>-cluster-ca-cert` secret. Yes.

**Reply by spoorthyredi2** [reply]
Ok.but that dint work for us.

I used the base64 decode of ca.crt  and it just dint work... we had to set NODE_TLS_REJECT_UNAUTHORIZED to false only then it worked

**Reply by scholzj** [reply]
You probably did something wrong then. Hard to say without knowing the exact steps or the details of the Javascript client I'm afraid 🤷

**Comment by archupsg03** [comment]
Am facing issues in java client itself. Can you help me sharing the steps to make strimzi work
